Add a new helper type for  SAFE.Client: `Optimistic`

aplikethewatch opened the pull request:
supports frontend development.

what changed?: add new helper client types for optimistic update
why?: aid in front end development

implements #11

aplikethewatch committed:
[feat]: Add new helper client types for optimistic update to support frontend development.
what changed?: add new helper client types for optimistic update
why?: aid in front end development
effect?: n/a

aplikethewatch committed:
[feat]: Updated XML docs; changed Curr property to Value for clarity.
what changed?: more xml docs and change Curr property to Value for more familiar usage
why?: doing x.Curr makes less sense than doing x.Value
effect?: n/a

Larocceau commented:
Interesting addition, curious to see it in action!

A few pieces of feedback:
* Can you add unit tests? I know the behaviour is trivial, but it's still nice to have unit tests so we can refactor more freely
* Can you add a module that has functions corresponding to the member functions on the type?
* I think it would be good to have some more basic functions for working with monads implemented, like map and bind, and perhaps we should add an asOption function as well. Look at the remoteData for more inspiration

aplikethewatch committed:
[feat]: Added tests and utility functions for Optimistic helper type based on PR feedback.
what changed?: added tests for Optimistic helper type; Also added utility functions for Optimistic helper type
why?: from pr feedback
effect?: better tested library

aplikethewatch committed:
[fix]: Opted for DU over record type to prevent illegal states.
what changed?: opt in for a DU instead of a record type for Optimistic helper type
why?: make illegal states unrepresentable
effect?: n/a

mattgallagher92 commented on src/SAFE.Client/SAFE.fs:
```diff
@@ -185,4 +185,53 @@ module RemoteData =
+
+///A type which represents optimistic updates. 
+type Optimistic<'T> =
+    | NonExistant
```
```suggestion
    | NonExistent
```

The correct spelling is existent

mattgallagher92 commented on src/SAFE.Client/SAFE.fs:
```diff
@@ -185,4 +185,53 @@ module RemoteData =
+    let startLoading (remote: RemoteData<'T>) = remote.StartLoading
+
+///A type which represents optimistic updates. 
+type Optimistic<'T> =
```
I feel like the case names could perhaps be clearer.

aplikethewatch replied:
Could it be represented as an option instead? like

```fsharp
type Optimistic<'T> = (current: 'T * previous: 'T option) option
```

mattgallagher92 commented on src/SAFE.Client/SAFE.fs:
```diff
@@ -185,4 +185,53 @@ module RemoteData =
+            | Exists (v, pv) -> Exists (value, Some v)
+
+        /// Rolls back to the previous value, discarding the current one.
+        member this.Rollback () =
```
```suggestion
        member this.RollBack () =
```

"Rollback" is a noun, "roll back" is the verb

aplikethewatch replied:
I honestly prefer `Rollback`

aplikethewatch committed:
Update src/SAFE.Client/SAFE.fs

Co-authored-by: Matt Gallagher <46973220+mattgallagher92@users.noreply.github.com>

Larocceau commented:
@arpxspace All actual code for this project has been moved to SAFE.Utils. Can you move this PR there?

aplikethewatch commented:
This PR has now moved [here](https://github.com/SAFE-Stack/SAFE.Utils/pull/4)
